profiling/src/profile/mod.rs

```diff
@@ -430,14 +437,76 @@ impl Profile {
         // size of 32KiB should definitely out-perform starting at zero for
         // time consumed, allocator pressure, and allocator fragmentation.
         const INITIAL_PPROF_BUFFER_SIZE: usize = 32 * 1024;
-        let mut buffer: Vec<u8> = Vec::with_capacity(INITIAL_PPROF_BUFFER_SIZE);
-        profile.encode(&mut buffer)?;
+        let mut encoder = CompressedProtobufSerializer::with_capacity(INITIAL_PPROF_BUFFER_SIZE);
+
+        for (sample, timestamp, mut values) in std::mem::take(&mut self.observations).into_iter() {
+            let labels = self.translate_and_enrich_sample_labels(sample, timestamp)?;
+            let location_ids: Vec<_> = self
+                .get_stacktrace(sample.stacktrace)
+                .locations
+                .iter()
+                .map(Id::to_raw_id)
+                .collect();
+            self.upscaling_rules
+                .upscale_values(&mut values, &labels, &self.sample_types)?;
+
+            let item = pprof::Sample {
+                location_ids,
+                values,
+                labels,
+            };
+
+            encoder.encode(ProfileSamplesEntry::from(item))?;
+        }
+
+        // `Sample`s must be emitted before `SampleTypes` since we use consume
+        // fields as we convert (using `into_iter`).  This allows Rust to
```

Comment:
Minor:

```suggestion
        // `Sample`s must be emitted before `SampleTypes` since we consume
        // fields as we convert (using `into_iter`).  This allows Rust to
```